src/Filecoin/Address.cpp

```diff
@@ -0,0 +1,169 @@
+// Copyright © 2017-2019 Trust.
+//
+// This file is part of Trust. The full Trust copyright notice, including
+// terms governing use, modification, and redistribution, is contained in the
+// file LICENSE at the root of the source code distribution tree.
+
+#include "Address.h"
+
+#include "../Base32.h"
+#include "../Data.h"
+
+using namespace TW;
+using namespace TW::Filecoin;
+
+static const char BASE32_ALPHABET_FILECOIN[] = "abcdefghijklmnopqrstuvwxyz234567";
+static constexpr size_t checksum_size = 4;
+
+bool Address::isValid(const Data& data) {
+    if (data.size() < 2) {
+        return false;
+    }
+    if (!get_type(data[0])) {
+        return false;
+    }
+    Type type = *get_type(data[0]);
+    if (type == Type::ID) {
+        // Verify varuint encoding
+        if (data.size() > 11)
+            return false;
+        if (data.size() == 11 && data[10] > 0x01)
+            return false;
+        int i;
+        for (i = 1; i < data.size(); i++)
+            if ((data[i] & 0x80) == 0)
+                break;
+        return i == data.size() - 1;
+    } else {
+        return data.size() == 1 + Address::payload_size(type);
+    }
+}
+
+static bool isValidID(const std::string &string) {
+    if (string.length() > 22)
+        return false;
+    try {
+        std::stoull(string.substr(2));
+    } catch (...) {
+        return false;
+    }
+    return true;
+}
+
+static bool isValidBase32(const std::string &string, Address::Type type) {
+    // Check if valid Base32.
+    uint8_t size = Address::payload_size(type);
+    Data decoded;
+    if (!Base32::decode(string.substr(2), decoded, BASE32_ALPHABET_FILECOIN)) {
+        return false;
+    }
+
+    // Check size
+    if (decoded.size() != size + 4) {
+        return false;
+    }
+
+    // Extract raw address.
+    Data address;
+    address.push_back(static_cast<uint8_t>(type));
+    address.insert(address.end(), decoded.data(), decoded.data() + size);
+
+    // Verify checksum.
+    Data should_sum = Hash::blake2b(address, checksum_size);
+    return std::memcmp(should_sum.data(), decoded.data() + size, checksum_size) == 0;
+}
+
+bool Address::isValid(const std::string& string) {
+    if (string.length() < 3) {
+        return false;
+    }
+    // Only test net addresses supported.
+    if (string[0] != 't') {
```

Comment:
can we add logic for mainnet? and remove test net

Reply:
Yep, confirmed with Filecoin that we are targetting mainnet instead now.